src/hotspot/share/opto/superword.cpp

```diff
@@ -2995,7 +2805,89 @@ bool SuperWord::output() {
         Node* one = vector_opd(p, 3);
         vn = VectorNode::make(opc, in, zero, one, vlen, velt_basic_type(n));
         vlen_in_bytes = vn->as_Vector()->length_in_bytes();
-      } else if (n->req() == 3 && !is_cmov_pack(p)) {
+      } else if (n->is_Cmp()) {
+        // Bool + Cmp + CMove -> VectorMaskCmp + VectorBlend
+        continue;
+      } else if (n->is_Bool()) {
+        // Bool + Cmp + CMove -> VectorMaskCmp + VectorBlend
+        continue;
+      } else if (n->is_CMove()) {
+        // Bool + Cmp + CMove -> VectorMaskCmp + VectorBlend
+
+        BoolNode* bol = n->in(1)->as_Bool();
+        assert(bol != nullptr, "must have Bool above CMove");
+        BoolTest::mask bol_test = bol->_test._test;
+        assert(bol_test == BoolTest::eq ||
+               bol_test == BoolTest::ne ||
+               bol_test == BoolTest::ge ||
+               bol_test == BoolTest::gt ||
+               bol_test == BoolTest::lt ||
+               bol_test == BoolTest::le,
+               "CMove bool should be one of: eq,ne,ge,ge,lt,le");
+        Node_List* p_bol = my_pack(bol);
+        assert(p_bol != nullptr, "CMove must have matching Bool pack");
+
+        CmpNode* cmp = bol->in(1)->as_Cmp();
+        assert(cmp != nullptr, "must have cmp above CMove");
+        Node_List* p_cmp = my_pack(cmp);
+        assert(p_cmp != nullptr, "Bool must have matching Cmp pack");
+
+        Node* cmp_in1 = vector_opd(p_cmp, 1);
+        Node* cmp_in2 = vector_opd(p_cmp, 2);
+
+        Node* blend_in1 = vector_opd(p, 2);
+        Node* blend_in2 = vector_opd(p, 3);
+
+        if (cmp->Opcode() == Op_CmpF || cmp->Opcode() == Op_CmpD) {
+          // If we have a Float or Double comparison, we must be careful with
+          // handling NaN's correctly. CmpF and CmpD have a return code, as
+          // they are based on the java bytecodes fcmpl/dcmpl:
+          // -1: cmp_in1 <  cmp_in2, or at least one of the two is a NaN
+          //  0: cmp_in1 == cmp_in2  (no NaN)
+          //  1: cmp_in1 >  cmp_in2  (no NaN)
+          //
+          // The "bol_test" selects which of the [-1, 0, 1] cases lead to "true".
+          //
+          // Note: ordered   (O) comparison returns "false" if either input is NaN.
+          //       unordered (U) comparison returns "true"  if either input is NaN.
+          //
+          // The VectorMaskCmpNode does a comparison directly on in1 and in2, in the java
+          // standard way (all comparisons are ordered, except NEQ is unordered).
```

Comment:
Sorry, I'm a bit confusing about the comment here. Based on your following description, are `LT` and `LE` unordered either?

Reply:
The issue is this:
`CmpF -> Bool [lt/le]` is unordered, because they both accept the return code `-1` from the `CmpF`, which also makes comparisons with `NaN` true. This means that such comparisons are `unordered`.

But `VectorMaskCmp` would interpret `lt/le` test-codes as `ordered`, so they would return false for `NaN` comparisons. So that is why we need to make a transformation here.

How would you improve my comments?

Reply:
@fg1417 thanks for the suggestion about running with the flags over all jtreg. I'll do that now...

Reply:
> The issue is this: `CmpF -> Bool [lt/le]` is unordered, because they both accept the return code `-1` from the `CmpF`, which also makes comparisons with `NaN` true. This means that such comparisons are `unordered`.
> 
> But `VectorMaskCmp` would interpret `lt/le` test-codes as `ordered`, so they would return false for `NaN` comparisons. So that is why we need to make a transformation here.
> 
> How would you improve my comments?

Thanks for your clarification.

Your comment is quite clear already. Maybe just highlight the mismatch between `VectorMaskCmp` and `bol_test` here, like:
```
          //
          // But with these two cases, which `VectorMaskCmp` interprets as ordered,
          // we must convert the unordered into an ordered comparison:
          //   BoolTest::lt:  Case -1    -> LT_U
          //   BoolTest::le:  Case -1, 0 -> LE_U
          //
```

Reply:
@fg1417 Are you ok with how I worded it now?

Reply:
> @fg1417 Are you ok with how I worded it now?

Oh, yes. Clear enough!

Reply:
@fg1417 Is there anything you still want me to change before you could approve this PR?

Reply:
> @fg1417 Is there anything you still want me to change before you could approve this PR?

Sorry for my delay. I assume that the flags `-XX:+UseCMoveUnconditionally` and `-XX:+UseVectorCmov` over all jtreg passed.

Reply:
Ah yes, they have passed indeed. With and without those flags, up to tier5 and stress testing :blush: